### PR TITLE
Introduce config(java system property) to start solr node with overseer role

### PR DESCRIPTION
**Pull request by noblepaul**
start a node with a system property 

`-Doverseer.node=true`
to set it as an overseer designate

to unset restart the node with 

`-Doverseer.node=false`

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 1]
```diff
@@ -2375,21 +2375,32 @@ public void rejoinShardLeaderElection(SolrParams params) {
...
+        }
       }
+      if (cmd == null) return;
+      ZkNodeProps props = new ZkNodeProps(Overseer.QUEUE_OPERATION, cmd.toLower(),
+          "node", getNodeName(),
+          "role", "overseer");
+      log.info("Going to add/remove role {} ", props);
+      getOverseerCollectionQueue().offer(Utils.toJSON(props));
     } catch (NoNodeException nne) {
       return;
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java]: If node is marked for overseer then for any exception should we just shutdown corecontainer?

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 2]
```diff
@@ -2375,21 +2375,32 @@ public void rejoinShardLeaderElection(SolrParams params) {
 
   public void checkOverseerDesignate() {
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:2395]: I saw that this function get called in coreContainer.load()? do we want to check this bit early (after creating a zkController?)

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 3]
```diff
@@ -2375,21 +2375,32 @@ public void rejoinShardLeaderElection(SolrParams params) {
...
+        if (overseerDesignates.contains(getNodeName())) cmd = CollectionParams.CollectionAction.ADDROLE;
+      } else {
+        if (isOverseer == Boolean.TRUE) {
+          cmd = CollectionParams.CollectionAction.ADDROLE;
+        } else if (isOverseer == Boolean.FALSE) {
+          if (overseerDesignates.contains(getNodeName())) cmd = CollectionParams.CollectionAction.REMOVEROLE;
+        }
       }
+      if (cmd == null) return;
+      ZkNodeProps props = new ZkNodeProps(Overseer.QUEUE_OPERATION, cmd.toLower(),
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:2410]: Minor: if node is already in overseer role and we restart the node, then do we need to "addRole" again? (or it is just safe to add that role again?)

Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:2410]: add-role is a command to overseer and it is idempotent

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 2 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java:2395]: this is the way it has always been. We wouldn't want to change the place

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java (hunk 1 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: yeah, that error will propagate up

Comment (hiteshk25) [reply solr/core/src/java/org/apache/solr/cloud/ZkController.java]: looks like we are eating exception here?
```
} catch (NoNodeException nne) {
      return;
    } catch (Exception e) {
      log.warn("could not read the overseer designate ", e);
    }
```

File: solr/core/src/java/org/apache/solr/cloud/ZkController.java [hunk 4]
```diff
@@ -2375,23 +2375,34 @@ public void rejoinShardLeaderElection(SolrParams params) {
...
+          if (overseerDesignates.contains(getNodeName())) cmd = CollectionParams.CollectionAction.REMOVEROLE;
+        }
+      }
+      if (cmd == null) return;
+      ZkNodeProps props = new ZkNodeProps(Overseer.QUEUE_OPERATION, cmd.toLower(),
+          "node", getNodeName(),
+          "role", "overseer");
+      log.info("Going to add/remove role {} ", props);
+      getOverseerCollectionQueue().offer(Utils.toJSON(props));
     } catch (Exception e) {
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/ZkController.java:2415]: @noblepaul should we remove this exception as well?

**Comment by hiteshk25** [comment]
Also, in createRepica/addReplica path add check if node role is "overseer" then don't create any replica? (another safety)

**Comment by noblepaul** [comment]
>Would be good to log warning here?

I could do that

**Comment by noblepaul** [comment]
>Also, in createRepica/addReplica path add check if node role is "overseer" then don't create any replica? (another safety)

everything go through the same code

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java [hunk 5]
```diff
@@ -573,6 +576,25 @@ public AssignRequest build() {
...
+          Map<String, Object> roles = (Map<String, Object>) Utils.fromJSON(data.getData());
+          List<String> dedicatedOverseers = (List<String>) roles.getOrDefault("overseer", Collections.emptyList());
+          for (String dedicatedOverseer : dedicatedOverseers) {
+            if(nodeList.contains(dedicatedOverseer)) {
+              nodeList.remove(dedicatedOverseer);
+            }
+          }
+        }
+
+      } catch (Exception e) {
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java]: @noblepaul @chatman I am not quite comfortable on eating the exception here and that other place above.  We have big clusters and things happen continuously(shard move, shard spilt, gc, zk connect/d) and we may run into this some day. Would be good to have some retry or fallback on this! 

Happy to chat more on this!

**Comment by hiteshk25** [comment]
> > Also, in createRepica/addReplica path add check if node role is "overseer" then don't create any replica? (another safety)
> 
> everything go through the same code

Some thoughts....

1. Is it possible to have an "overseer-role node" check, before creating a 'solrcore' instance? Probably that will ensure we will not create any replica on that overseer-designated node?
2.Or ClusterState.getLiveNodes() : if this API just returns live solr data nodes(or filters overseer designated nodes)? But there may be some race conditions if the overseer role is created later. (at client level we are differentiating solr nodes with port but that will be hacky! )

Happy to chat more on this. Ping me on slack!

@noblepaul @chatman

**Comment by chatman** [comment]
Either of the two solutions are reasonable and helpful. Such safeguards really complete the "dedicated overseers" story from Solr's point of view, i.e. it then feels like a first class support.

**Comment by noblepaul** [comment]
>Or ClusterState.getLiveNodes() : 


This a dangerous change. This code is used everywhere

**Comment by noblepaul** [comment]
@hiteshk25 

We now prevent creation of a core in an overseer designate, please review

**Comment by hiteshk25** [comment]
I have run the solr tests locally. Following tests are failing
```
    > Caused by: org.apache.solr.common.SolrException: org.apache.solr.common.SolrException: java.lang.NullPointerException
16624    [junit4]    >    at org.apache.solr.cloud.autoscaling.sim.SimCloudManager.simHandleSolrRequest(SimCloudManager.java:905)
16625    [junit4]    >    at org.apache.solr.cloud.autoscaling.sim.SimCloudManager.lambda$request$15(SimCloudManager.java:734)
16626    [junit4]    >    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
16627    [junit4]    >    at org.apache.solr.common.util.ExecutorUtil$MDCAwareThreadPoolExecutor.lambda$execute$0(ExecutorUtil.java:218)
16628    [junit4]    >    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
16629    [junit4]    >    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
16630    [junit4]    >    ... 1 more
16631    [junit4]    > Caused by: org.apache.solr.common.SolrException: java.lang.NullPointerException
16632    [junit4]    >    at org.apache.solr.cloud.ZkController.getDedicatedOverseers(ZkController.java:2390)
16633    [junit4]    >    at org.apache.solr.cloud.api.collections.Assign$LegacyAssignStrategy.assign(Assign.java:563)
16634    [junit4]    >    at org.apache.solr.cloud.api.collections.CreateCollectionCmd.buildReplicaPositions(CreateCollectionCmd.java:478)
16635    [junit4]    >    at org.apache.solr.cloud.autoscaling.sim.SimClusterStateProvider.simCreateCollection(SimClusterStateProvider.java:976)
16636    [junit4]    >    at org.apache.solr.cloud.autoscaling.sim.SimCloudManager.simHandleSolrRequest(SimCloudManager.java:903)
16637    [junit4]    >    ... 6 more

```

```
26396    [junit4]
26397    [junit4] Tests with failures [seed: F74B5E0ED5784847] (first 10 out of 34):
26398    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimScenario.testIndexing
26399    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimScenario.testSplitShard
26400    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimScenario.testSuggestions
26401    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimLargeCluster.testCreateLargeSimCollections
26402    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimLargeCluster.testNodeLost
26403    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimLargeCluster.testAddNode
26404    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimLargeCluster.testSearchRate
26405    [junit4]   - org.apache.solr.cloud.autoscaling.sim.TestSimLargeCluster.testFreediskTracking
26406    [junit4]   - org.apache.solr.handler.admin.MetricsHistoryHandlerTest (suite)
26407    [junit4]   - org.apache.solr.cloud.autoscaling.ScheduledTriggerIntegrationTest.testScheduledTrigger

```

**Comment by noblepaul** [comment]
it'a an autoscaling test. it uses a simulator and not a real test

**Review by hiteshk25 — APPROVED** [review]
LGTM Thanks @noblepaul

**Pull request merged**